safekeeper/src/wal_backup.rs

```diff
@@ -419,73 +419,37 @@ static REMOTE_STORAGE: OnceCell<Option<GenericRemoteStorage>> = OnceCell::new();
 async fn backup_object(source_file: &Path, size: usize) -> Result<()> {
     let storage = REMOTE_STORAGE.get().expect("failed to get remote storage");
 
-    let file = File::open(&source_file).await?;
-
-    // Storage is initialized by launcher at this point.
-    match storage.as_ref().unwrap() {
-        GenericRemoteStorage::Local(local_storage) => {
-            let destination = local_storage.remote_object_id(source_file)?;
-
-            debug!(
-                "local upload about to start from {} to {}",
-                source_file.display(),
-                destination.display()
-            );
-            local_storage.upload(file, size, &destination, None).await
-        }
-        GenericRemoteStorage::S3(s3_storage) => {
-            let s3key = s3_storage.remote_object_id(source_file)?;
-
-            debug!(
-                "S3 upload about to start from {} to {:?}",
-                source_file.display(),
-                s3key
-            );
-            s3_storage.upload(file, size, &s3key, None).await
-        }
-    }?;
+    let file = tokio::io::BufReader::new(File::open(&source_file).await.with_context(|| {
+        format!(
+            "Failed to open file {} for wal backup",
+            source_file.display()
+        )
+    })?);
 
-    Ok(())
+    storage
+        .as_ref()
+        .expect("Storage should be initialized by launcher at this point.")
+        .upload_storage_object(file, size, source_file)
+        .await
 }
 
 pub async fn read_object(
     file_path: PathBuf,
     offset: u64,
 ) -> anyhow::Result<Pin<Box<dyn tokio::io::AsyncRead>>> {
-    let download = match REMOTE_STORAGE
+    let download = REMOTE_STORAGE
         .get()
         .context("Failed to get remote storage")?
         .as_ref()
         .context("No remote storage configured")?
-    {
-        GenericRemoteStorage::Local(local_storage) => {
-            let source = local_storage.remote_object_id(&file_path)?;
-
-            info!(
-                "local download about to start from {} at offset {}",
-                source.display(),
-                offset
-            );
-            local_storage
-                .download_byte_range(&source, offset, None)
-                .await
-        }
-        GenericRemoteStorage::S3(s3_storage) => {
-            let s3key = s3_storage.remote_object_id(&file_path)?;
-
-            info!(
```

Comment:
I'd like to keep this log, in case something will go wrong with S3 download. Download can't be triggered more often than once per 16MB, so these logs shouldn't be a spam problem.

Reply:
https://github.com/neondatabase/neon/pull/2381